### PR TITLE
Add more documentation to SortPreservingMergeStream

### PR DESCRIPTION
**Pull request by aprimadi**
# Which issue does this PR close?

N/A

# Rationale for this change

Just adding some documentation to help future readers.

# What changes are included in this PR?

More documentation + inline functions

# Are these changes tested?

Yes

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by aprimadi** [comment]
Adding a test just to convince myself that the loser tree implementation work even with unbalanced / incomplete binary tree.

**Comment by aprimadi** [comment]
Perhaps I've overdone this

**Comment by alamb** [comment]
> Perhaps I've overdone this

No I think this is great -- thank you very much.

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 1]
```diff
@@ -506,6 +506,71 @@ mod tests {
         assert_batches_eq!(exp, collected.as_slice());
     }
 
+    #[tokio::test]
+    async fn test_merge_five_partitions() {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:510]: I wonder if you can leave some comments about why this test was added / what it is covering to help future readers?

I think the coverage of merging is pretty good with this: https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/merge_fuzz.rs

**Review by alamb — APPROVED** [review]
Thank you so much for this

cc @yjshen and @richox and @tustvold in case you would like to add to the discussion or have more

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs (hunk 1 above)
Comment (aprimadi) [reply datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:510]: Ah yes, I added that test just to convince myself that the implementation of loser tree works.

Now I'm pretty sure it works, perhaps we can remove it?

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 2]
```diff
@@ -506,6 +506,71 @@ mod tests {
...
+            "| 11 |",
+            "| 13 |",
+            "| 13 |",
+            "| 17 |",
+            "+----+",
+        ];
+        let collected = collect(merge, task_ctx).await.unwrap();
+        assert_batches_eq!(expected, collected.as_slice());
+    }
+
```
Comment (aprimadi) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:573]:
```suggestion
```

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:510]: I leave that to your judgement

**Pull request merged**